channeldb/graph_cache.go

```diff
@@ -0,0 +1,330 @@
+package channeldb
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcutil"
+
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+// GraphNode is an interface for all the information the cache needs to know
+// about a lightning node.
+type GraphNode interface {
+	// PubKey is the node's public identity key.
+	PubKey() route.Vertex
+
+	// Features returns the node's p2p features.
+	Features() *lnwire.FeatureVector
+
+	// ForEachChannel iterates through all channels of a given node,
+	// executing the passed callback with an edge info structure and the
+	// policies of each end of the channel. The first edge policy is the
+	// outgoing edge *to* the connecting node, while the second is the
+	// incoming edge *from* the connecting node. If the callback returns an
+	// error, then the iteration is halted with the error propagated back up
+	// to the caller.
+	ForEachChannel(kvdb.RTx,
+		func(kvdb.RTx, *ChannelEdgeInfo, *ChannelEdgePolicy,
+			*ChannelEdgePolicy) error) error
+}
+
+// DirectedChannel is a type that stores the channel information as seen from
+// one side of the channel.
+type DirectedChannel struct {
+	// ChannelID is the unique identifier of this channel.
+	ChannelID uint64
+
+	// IsNode1 indicates if this is the node with the smaller public key.
+	IsNode1 bool
+
+	// OtherNode is the public key of the node on the other end of this
+	// channel.
+	OtherNode route.Vertex
+
+	// Capacity is the announced capacity of this channel in satoshis.
+	Capacity btcutil.Amount
+
+	// OutPolicy is the outgoing policy from this node *to* the other node.
+	OutPolicy *ChannelEdgePolicy
+
+	// InPolicy is the incoming policy *from* the other node to this node.
+	InPolicy *ChannelEdgePolicy
+}
+
+// GraphCache is a type that holds a minimal set of information of the public
+// channel graph that can be used for pathfinding.
+type GraphCache struct {
+	nodeChannels map[route.Vertex][]*DirectedChannel
+	nodeFeatures map[route.Vertex]*lnwire.FeatureVector
+
+	mtx sync.RWMutex
+}
```

Comment:
So no synergy with the existing channel cache? Does keep the code simpler here without too many fancy pointing structures.

Reply:
Curious w.r.t this as well, given that every time an edge gets updated it's added to the cache: https://github.com/lightningnetwork/lnd/blob/master/channeldb/graph.go

Also, if a single peer ends up requesting us to send them an update backlog, we'll end up loading a potentially large amount into memory: https://github.com/lightningnetwork/lnd/blob/master/channeldb/graph.go#L1592

Reply:
The main problem with the existing channel cache is that it is limited in size and items are removed generously because we know we'll get them cached again later.
For the graph cache, we need to be sure that it is always complete and fully up to date. Otherwise we'll get bad results when trying to find paths.

I think we could maybe make use of the graph cache to get rid of the channel cache in the future. But I'd rather not do that in this PR as well.

Reply:
> The main problem with the existing channel cache is that it is limited in size and items are removed generously because we know we'll get them cached again later

Ahh makes sense, that's a shame, though we'll only really reach the full extent of double caching if a peers sends us a non-zero timestamp for "chans in horizon".